### PR TITLE
feat(agnocastlib, ros2agnocast_discovery_agent): add cross-NS daemon bridge MQ and let the discovery agent dispatch bridge requests

### PR DESCRIPTION
**Pull request by k1832**
## Description

Auto-generate bridges between Agnocast endpoints that live in different **IPC namespaces** (or ECUs). This PR adds the shared plumbing plus the discovery agent that drives it (how it decides whether to make a request and how it sends it); the per-NS `bridge_manager` that acts on the requests is in the stacked PR.

Each per-NS discovery agent compares its local Agnocast state against the remote gossip snapshots. When a topic has a local endpoint whose opposite-role peer lives in another namespace, the agent asks the `bridge_manager` in its namespace to build the matching half of the relay:

- local **publisher** + remote **subscriber** → an **A2R** bridge (Agnocast → DDS)
- local **subscriber** + remote **publisher** → an **R2A** bridge (DDS → Agnocast)

Each namespace builds only its own half; the halves connect over DDS (e.g. NS-A's A2R → DDS → NS-B's R2A), so the two endpoints reach each other.

Changes:

- **agnocastlib**: a new agent→manager request `MqMsgDaemonBridge` on its own MQ (`create_mq_name_for_daemon_bridge`), which the bridge event loop can also listen on (`register_aux_mq`).
- **ros2agnocast_discovery_agent**: `bridge_decider` computes the needed requests each tick and dispatches `MqMsgDaemonBridge` to the per-NS `bridge_manager` MQ.

**Scope:** the request path targets the per-NS (performance) `bridge_manager`. The standard `bridge_manager` is intentionally not wired: the standard bridge is planned to be unified into the per-NS (performance) bridge. See the design link below.

This reuses the existing factory mechanism; additive only — no existing ABI changes, kmod untouched.

## Related links

- Cross-namespace bridge design (TIER IV internal link): https://tier4.atlassian.net/wiki/x/2YFtNwE
- Architecture overview (TIER IV internal link): https://tier4.atlassian.net/wiki/x/fIDcMwE
- Preceded by #1388 (bridge `request`→`registration` rename, merged into main)

## How was this PR tested?

- [ ] Autoware
- [x] `bash scripts/test/e2e_test_1to1.bash` (required)
- [x] `bash scripts/test/e2e_test_2to2.bash` (required)
- [ ] kunit tests (required when modifying the kernel module) — N/A, kmod untouched
- [x] `bash scripts/test/run_requires_kernel_module_tests.bash` (required)
- [ ] sample application

In addition:

- **Unit tests**: `bridge_decider` pytest (direction/role logic, dedup, type fallback, wire format, per-NS MQ dispatch) and a C++ test pinning the `MqMsgDaemonBridge` wire layout (`sizeof == 524`, field offsets) to the agent's Python `struct` format. `agnocastlib` gtest and the discovery-agent pytest pass.
- The end-to-end cross-NS behavior is exercised together with the stacked performance-manager PR.

## Version Update Label (Required)

- `need-patch-update`

### CODE REVIEW

File: src/ros2agnocast_discovery_agent/ros2agnocast_discovery_agent/bridge_decider.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                continue
+
+            local_pubs = [p for p in local_topic.publishers if not p.is_bridge]
+            local_subs = [s for s in local_topic.subscribers if not s.is_bridge]
+            remote_pubs = [p for p in remote_topic.publishers if not p.is_bridge]
+            remote_subs = [s for s in remote_topic.subscribers if not s.is_bridge]
+
+            type_name = local_topic.type_name or remote_topic.type_name
+            if not type_name:
+                continue
```
[the remark this answers is not in the document]
Comment (k1832) [reply src/ros2agnocast_discovery_agent/ros2agnocast_discovery_agent/bridge_decider.py:122]: Good catch — fixed. Type is now resolved per-topic across local + all remotes (`_resolve_types`) before the decision loop, so a remote that supplies the opposite-role endpoint but lacks the type no longer causes a skip when another snapshot for the same topic has it.

File: src/ros2agnocast_discovery_agent/test/test_bridge_decider.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+def test_decide_uses_remote_type_when_local_missing():
+    local = _state(topics=[_topic('/x', type_name='', pubs=[_endpoint('/lp')])])
+    remote = _state(host_uuid='OTHER', ipc_ns=222,
+                    topics=[_topic('/x', type_name='std_msgs/msg/String',
+                                   subs=[_endpoint('/rs')])])
+    reqs = decide_bridges(local, {('OTHER', 222): remote})
+    assert len(reqs) == 1
+    assert reqs[0].type_name == 'std_msgs/msg/String'
+
+
```
[the remark this answers is not in the document]
Comment (k1832) [reply src/ros2agnocast_discovery_agent/test/test_bridge_decider.py:122]: Added `test_decide_uses_type_from_other_remote_when_matching_remote_missing` covering exactly this: the opposite-role remote has an empty type while another remote provides it.

File: src/agnocastlib/test/unit/test_daemon_bridge_mq.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+TEST(DaemonBridgeMqTest, PerformanceMqNameAppendsDomainId)
+{
+  setenv("ROS_DOMAIN_ID", "7", 1);
+  EXPECT_EQ(
+    agnocast::create_mq_name_for_daemon_bridge(agnocast::PERFORMANCE_BRIDGE_VIRTUAL_PID),
+    "/agnocast_daemon_bridge_perf_d7");
+  unsetenv("ROS_DOMAIN_ID");
+}
```
[the remark this answers is not in the document]
Comment (k1832) [reply src/agnocastlib/test/unit/test_daemon_bridge_mq.cpp:76]: Fixed — added a `ScopedRosDomainId` RAII guard that captures the pre-test value in its ctor and restores (or unsets) it in its dtor, so these tests no longer leak `ROS_DOMAIN_ID` into later tests.

File: src/agnocastlib/test/unit/test_daemon_bridge_mq.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
+#include "agnocast/agnocast_mq.hpp"
+#include "agnocast/agnocast_utils.hpp"
+
+#include <gtest/gtest.h>
+
+#include <cstddef>
+#include <cstdlib>
+
```
[the remark this answers is not in the document]
Comment (k1832) [reply src/agnocastlib/test/unit/test_daemon_bridge_mq.cpp:39]: Added `#include <string>` — the restore guard stores the previous value as `std::string`.

File: src/ros2agnocast_discovery_agent/ros2agnocast_discovery_agent/bridge_decider.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                sub = local_subs[0]
+                key = (local_topic.topic_name, DIRECTION_ROS2_TO_AGNOCAST)
+                requests.setdefault(key, BridgeRequest(
+                    topic_name=local_topic.topic_name,
+                    type_name=type_name,
+                    direction=DIRECTION_ROS2_TO_AGNOCAST,
+                    qos_depth=sub.qos_depth,
+                    qos_is_transient_local=sub.qos_is_transient_local,
+                    qos_is_reliable=sub.qos_is_reliable,
+                ))
```
[the remark this answers is not in the document]
Comment (k1832) [reply src/ros2agnocast_discovery_agent/ros2agnocast_discovery_agent/bridge_decider.py:146]: Agreed that picking `[0]` is endpoint-order dependent. A deterministic aggregate QoS (e.g. max depth + strongest durability/reliability) is a slightly larger design choice, so I'll track it in a follow-up issue rather than expand this PR.

[the remark this answers is not in the document]
Comment (k1832) [reply src/ros2agnocast_discovery_agent/ros2agnocast_discovery_agent/bridge_decider.py:146]: Also, I guess it's not very common that the topic has multiple QoS?
cc: @Koichi98

**Comment by Koichi98** [comment]
@k1832 
Ideally I'd like to avoid adding a new aux MQ and instead reuse the existing MQ so the agent can `register` through it too. That requires refactoring what we put on the MQ (e.g. carrying QoS inline, dropping `target_id`), so it's fine to leave out of scope for this PR — let's do it after the standard/performance bridge unification.

File: src/agnocastlib/src/agnocast_utils.cpp [hunk 6]
```diff
@@ -91,6 +91,20 @@ std::string create_mq_name_for_bridge(const pid_t pid)
   return name;
 }
 
+std::string create_mq_name_for_daemon_bridge(const pid_t pid)
```
Comment (Koichi98) [inline src/agnocastlib/src/agnocast_utils.cpp:103]: `create_mq_name_for_daemon_bridge` treats an empty `ROS_DOMAIN_ID` (set but `""`) as "no domain" (`!domain_id.empty()`), but the existing `create_mq_name_for_bridge` only null-checks, so it appends a bare `_d` in that case.

The behavior of this `create_mq_name_for_daemon_bridge` is the correct one. Could we align the existing `create_mq_name_for_bridge` to the same empty-check in this PR (ideally factoring the suffix logic into a shared helper) so the two stay consistent?

Comment (Koichi98) [reply src/agnocastlib/src/agnocast_utils.cpp:103]: Same for [this](https://github.com/autowarefoundation/agnocast/pull/1389/changes#diff-2c6d3f48fcebadf0277d73648f5e341e5591f1d7144f32a829508dfeeb2459b0R151) too.

File: src/ros2agnocast_discovery_agent/test/test_bridge_decider.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+def _state(host_uuid='HOST', ipc_ns=111, topics=None):
+    s = AgnocastDaemonState()
+    s.schema_version = 1
+    s.host_uuid = host_uuid
+    s.ipc_ns_inode = ipc_ns
+    s.topics = topics or []
+    return s
+
```
Comment (Koichi98) [inline src/ros2agnocast_discovery_agent/test/test_bridge_decider.py:52]: There's no negative case for same-topic / same-role-only — e.g. local pub + remote pub (no subscriber anywhere) should emit nothing. It might be worth adding a test that asserts an empty result when the topic matches but the opposite-role peer is absent.

File: src/agnocastlib/src/agnocast_utils.cpp (hunk 6 above)
Comment (k1832) [reply src/agnocastlib/src/agnocast_utils.cpp:103]: Fixed — extracted a shared `performance_domain_suffix()` helper (treats unset *or* empty `ROS_DOMAIN_ID` as "no domain") and used it in both `create_mq_name_for_bridge` and `create_mq_name_for_daemon_bridge`, so they can't drift. The Python `_performance_mq_name` already used the empty-check, so all three now agree. Added a `PerformanceMqNameEmptyDomainIdHasNoSuffix` test pinning the empty case for both builders.

File: src/ros2agnocast_discovery_agent/test/test_bridge_decider.py (hunk 7 above)
Comment (k1832) [reply src/ros2agnocast_discovery_agent/test/test_bridge_decider.py:52]: Added `test_decide_skips_when_only_same_role_present`: same topic on both sides but publishers only (no subscriber anywhere) -> asserts an empty result.

**Comment by k1832** [comment]
Agreed — let's keep the aux MQ for now and fold the agent into the existing MQ when we refactor the MQ payload (QoS inline, drop `target_id`) as part of the standard/performance bridge unification. I'll track it there rather than in this PR.

**Review by Koichi98 — DISMISSED** [review]

**Review by Koichi98 — APPROVED** [review]
